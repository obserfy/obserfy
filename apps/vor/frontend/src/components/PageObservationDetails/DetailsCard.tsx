import { t, Trans } from "@lingui/macro"
import { FC, useState } from "react"
import { Button, Card, Flex } from "theme-ui"
import { borderBottom } from "../../border"
import usePatchObservation from "../../hooks/api/observations/usePatchObservation"
import useVisibilityState from "../../hooks/useVisibilityState"
import { ReactComponent as CloseIcon } from "../../icons/close.svg"
import Icon from "../Icon/Icon"
import MarkdownEditor from "../MarkdownEditor/MarkdownEditor"
import MultilineDataBox from "../MultilineDataBox/MultilineDataBox"
import { Typography } from "../Typography/Typography"

interface Props {
  originalValue: string
  observationId: string
}
const DetailsCard: FC<Props> = ({ originalValue, observationId }) => (
  <LongTextDataBox
    originalValue={originalValue}
    observationId={observationId}
  />
)

const LongTextDataBox: FC<{
  originalValue?: string
  observationId: string
}> = ({ originalValue, observationId }) => {
  const editor = useVisibilityState()

  if (editor.visible) {
    return (
      <DetailEditor
        onDismiss={editor.hide}
        onSave={editor.hide}
        observationId={observationId}
        initialValue={originalValue}
      />
    )
  }

  return (
    <Card variant="responsive" mb={3} pb={2}>
      <MultilineDataBox
        label={t`Details`}
        value={originalValue ?? ""}
        onEditClick={editor.show}
        placeholder="-"
      />
    </Card>
  )
}

const DetailEditor: FC<{
  initialValue?: string
  onDismiss: () => void
  onSave: () => void
  observationId: string
}> = ({ observationId, onDismiss, onSave, initialValue = "" }) => {
  const [value, setValue] = useState(initialValue)
  const patchObservation = usePatchObservation(observationId)

  const handleSave = async () => {
    try {
      await patchObservation.mutateAsync({ longDesc: value })
      onSave()
    } catch (e) {
      Sentry.captureException(e)
    }
  }

  return (
    <Card variant="responsive" mb={3} sx={{ overflow: "hidden" }}>
      <Flex pt={3} pb={3} sx={{ ...borderBottom, alignItems: "center" }}>
        <Typography.Body px={3} sx={{ fontWeight: "bold" }}>
          <Trans>Details</Trans>
        </Typography.Body>

        <Button variant="outline" ml="auto" p={1} mr={2} onClick={onDismiss}>
          <Icon as={CloseIcon} fill="danger" />
        </Button>

        <Button
          mr={3}
          sx={{ fontWeight: "bold", fontSize: 0 }}
          onClick={handleSave}
        >
          <Trans>Save</Trans>
        </Button>
      </Flex>

      <MarkdownEditor value={value} onChange={setValue} />
    </Card>
  )
}

export default DetailsCard
