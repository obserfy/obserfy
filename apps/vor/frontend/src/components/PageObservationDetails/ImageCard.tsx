import { Image, Card, Flex, Label } from "theme-ui"
import { FC, Fragment, useState } from "react"
import { Trans } from "@lingui/macro"
import Typography from "../Typography/Typography"
import usePostNewObservationImage from "../../hooks/api/observations/usePostNewObservationImage"
import { borderFull } from "../../border"
import LoadingIndicator from "../LoadingIndicator/LoadingIndicator"
import Icon from "../Icon/Icon"
import { ReactComponent as PlusIcon } from "../../icons/plus.svg"
import ImagePreviewOverlay from "../ImagePreviewOverlay/ImagePreviewOverlay"

interface ImageData {
  id: string
  thumbnailUrl: string
  originalUrl: string
}

interface Props {
  studentId: string
  observationId: string
  originalValue: ImageData[]
}
const ImageCard: FC<Props> = ({ originalValue, studentId, observationId }) => (
  <Card sx={{ borderRadius: [0, "default"] }} mx={[0, 3]} pt={3}>
    <Typography.Body
      mb={2}
      color="textMediumEmphasis"
      sx={{ lineHeight: 1, fontSize: [1, 0] }}
      px={3}
    >
      <Trans>Images</Trans>
    </Typography.Body>
    <ImagesDataBox
      studentId={studentId}
      observationId={observationId}
      images={originalValue}
    />
  </Card>
)

const ImagesDataBox: FC<{
  studentId: string
  images: ImageData[]
  observationId: string
}> = ({ studentId, images, observationId }) => {
  const [selectedIdx, setSelectedIdx] = useState<number>()
  const postNewImage = usePostNewObservationImage(observationId)
  const selectedImage =
    selectedIdx !== undefined ? images[selectedIdx] : undefined

  return (
    <Fragment>
      <Flex px={3} pb={2} sx={{ flexWrap: "wrap", alignItems: "center" }}>
        {images.map((image, idx) => {
          const isSelected = selectedIdx === idx
          return (
            <Image
              data-cy="image"
              key={image.id}
              src={image.thumbnailUrl}
              mr={2}
              mb={2}
              sx={{
                height: "40px",
                width: "40px",
                borderRadius: "default",
                cursor: "pointer",
                boxShadow: isSelected ? "0 0 0 2px #00a875" : "",
                ...borderFull,
              }}
              onClick={() => setSelectedIdx(isSelected ? undefined : idx)}
            />
          )
        })}
        <Label
          mb={2}
          sx={{
            height: "40px",
            width: "40px",
            backgroundColor: "background",
            borderRadius: "default",
            cursor: "pointer",
            "&:hover": { borderColor: "primary" },
            opacity: postNewImage.isLoading ? 0.5 : 1,
            ...borderFull,
          }}
          variant="outline"
          p={0}
        >
          <input
            data-cy="upload-image"
            type="file"
            sx={{ display: "none" }}
            accept="image/*"
            disabled={postNewImage.isLoading}
            onChange={async (event) => {
              if (event.target.files) {
                try {
                  await postNewImage.mutateAsync(event.target.files[0])
                } catch (e) {
                  Sentry.captureException(e)
                }
              }
            }}
          />
          {postNewImage.isLoading ? (
            <LoadingIndicator m="auto" />
          ) : (
            <Icon as={PlusIcon} m="auto" />
          )}
        </Label>
      </Flex>
      {selectedIdx !== undefined && (
        <ImagePreviewOverlay
          imageId={selectedImage?.id ?? ""}
          studentId={studentId}
          src={selectedImage?.originalUrl ?? ""}
          onDismiss={() => setSelectedIdx(undefined)}
        />
      )}
    </Fragment>
  )
}

export default ImageCard
