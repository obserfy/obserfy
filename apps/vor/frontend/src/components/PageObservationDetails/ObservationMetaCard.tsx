import { FC, useState } from "react"
import { Card, Flex } from "theme-ui"
import { t } from "@lingui/macro"
import dayjs, { Dayjs } from "../../dayjs"
import useGetObservation from "../../hooks/api/observations/useGetObservation"
import usePatchObservation from "../../hooks/api/observations/usePatchObservation"
import DataBox from "../DataBox/DataBox"
import Dialog from "../Dialog/Dialog"
import DialogHeader from "../DialogHeader/DialogHeader"
import Input from "../Input/Input"
import { useGetCurriculumAreas } from "../../hooks/api/useGetCurriculumAreas"
import useVisibilityState from "../../hooks/useVisibilityState"
import Chip from "../Chip/Chip"
import DatePickerDialog from "../DatePickerDialog/DatePickerDialog"

interface Props {
  observationId: string
}
const ObservationMetaCard: FC<Props> = ({ observationId }) => {
  const { data } = useGetObservation(observationId)

  return (
    <Card sx={{ borderRadius: [0, "default"] }} mx={[0, 3]} mb={3}>
      <ShortDescription
        originalValue={data?.shortDesc}
        observationId={observationId}
      />
      <EventTime
        originalValue={dayjs(data?.eventTime)}
        observationId={observationId}
      />
      <Area originalValue={data?.area?.id} observationId={observationId} />
    </Card>
  )
}

const ShortDescription: FC<{
  originalValue?: string
  observationId: string
}> = ({ originalValue, observationId }) => {
  const patchObservation = usePatchObservation(observationId)
  const [isEditing, setIsEditing] = useState(false)
  const [shortDesc, setShortDesc] = useState(originalValue)

  return (
    <>
      <DataBox
        label={t`Short Description`}
        value={originalValue ?? ""}
        onEditClick={() => setIsEditing(true)}
      />
      {isEditing && (
        <Dialog>
          <DialogHeader
            title={t`Edit Short Description`}
            onAcceptText={t`Save`}
            onCancel={() => setIsEditing(false)}
            loading={patchObservation.isLoading}
            onAccept={async () => {
              try {
                await patchObservation.mutateAsync({ shortDesc })
                setIsEditing(false)
              } catch (e) {
                Sentry.captureException(e)
              }
            }}
          />
          <Input
            label={t`Short Description`}
            sx={{ width: "100%" }}
            onChange={(e) => setShortDesc(e.target.value)}
            value={shortDesc}
            containerSx={{ p: 3, backgroundColor: "background" }}
          />
        </Dialog>
      )}
    </>
  )
}

const Area: FC<{
  originalValue?: string
  observationId: string
}> = ({ originalValue, observationId }) => {
  const { data: areas } = useGetCurriculumAreas()
  const patchObservation = usePatchObservation(observationId)
  const dialog = useVisibilityState()
  const [areaId, setValue] = useState(originalValue ?? "")

  const original = areas?.find((v) => v.id === originalValue)

  return (
    <>
      <DataBox
        label={t`Area`}
        value={original?.name ?? "-"}
        onEditClick={dialog.show}
      />
      {dialog.visible && (
        <Dialog>
          <DialogHeader
            title={t`Edit Area`}
            onAcceptText={t`Save`}
            onCancel={dialog.hide}
            loading={patchObservation.isLoading}
            onAccept={async () => {
              try {
                await patchObservation.mutateAsync({ areaId })
                dialog.hide()
              } catch (e) {
                Sentry.captureException(e)
              }
            }}
          />
          <Flex
            px={3}
            pt={3}
            pb={2}
            sx={{ flexWrap: "wrap", backgroundColor: "background" }}
          >
            {areas?.map(({ name, id }) => {
              const isSelected = id === areaId
              return (
                <Chip
                  key={id}
                  text={name}
                  activeBackground="primary"
                  isActive={isSelected}
                  onClick={() => setValue(isSelected ? "" : id)}
                  mr={2}
                  mb={2}
                />
              )
            })}
          </Flex>
        </Dialog>
      )}
    </>
  )
}

const EventTime: FC<{
  originalValue?: Dayjs
  observationId: string
}> = ({ observationId, originalValue }) => {
  const patchObservation = usePatchObservation(observationId)
  const dialog = useVisibilityState()

  return (
    <>
      <DataBox
        label={t`Event Time`}
        value={originalValue?.format("dddd, DD MMM YYYY") ?? ""}
        onEditClick={dialog.show}
      />
      {dialog.visible && (
        <DatePickerDialog
          isLoading={patchObservation.isLoading}
          defaultDate={originalValue}
          onDismiss={dialog.hide}
          onConfirm={async (eventTime) => {
            try {
              await patchObservation.mutateAsync({ eventTime })
              dialog.hide()
            } catch (e) {
              Sentry.captureException(e)
            }
          }}
        />
      )}
    </>
  )
}

export default ObservationMetaCard
