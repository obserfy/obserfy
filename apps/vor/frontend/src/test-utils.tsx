/* eslint-disable import/no-extraneous-dependencies */
import { FC, ReactElement } from "react"
import { QueryClient, QueryClientProvider } from "react-query"
import { ThemeProvider, Box, Theme } from "theme-ui"
import { render, RenderOptions, RenderResult } from "@testing-library/react"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router"
import { I18nProvider } from "@lingui/react"
import { i18n } from "@lingui/core"
import { en } from "make-plural/plurals"
import theme from "./gatsby-plugin-theme-ui"
// @ts-ignore
import enCatalog from "../i18n/lingui/en/messages.js"

// any used here because the catalogs are generated b lingui
// not our responsibility
i18n.loadLocaleData("en", { plurals: en })
i18n.load("en", enCatalog.messages as any)
i18n.activate("en")

const queryClient = new QueryClient()
const history = createHistory(createMemorySource("/"))

const AllProviders: FC = ({ children }) => {
  return (
    <I18nProvider i18n={i18n}>
      <QueryClientProvider client={queryClient}>
        <LocationProvider history={history}>
          <ThemeProvider theme={theme as Theme}>
            <Box sx={{ fontSize: [16, 20] }}>{children}</Box>
          </ThemeProvider>
        </LocationProvider>
      </QueryClientProvider>
    </I18nProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: AllProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
