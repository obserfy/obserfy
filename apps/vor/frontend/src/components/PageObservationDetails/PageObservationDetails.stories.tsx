import { Meta, Story } from "@storybook/react"
import PageObservationDetails, {
  PageObservationDetailsProps,
} from "./PageObservationDetails"

export default {
  title: "Core/PageObservationDetails",
  component: PageObservationDetails,
  parameters: {
    componentSubtitle: "Just a simple PageObservationDetails",
  },
} as Meta

const Template: Story<PageObservationDetailsProps> = () => (
  <PageObservationDetails observationId="" backUrl="/" studentId="" />
)

export const Default = Template.bind({})
Default.args = {}
