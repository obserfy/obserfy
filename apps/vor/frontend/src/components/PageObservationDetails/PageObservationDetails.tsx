import { FC, useState } from "react"
import { Box, Button, Flex } from "theme-ui"
import { t } from "@lingui/macro"
import useGetObservation from "../../hooks/api/observations/useGetObservation"
import LoadingPlaceholder from "../LoadingPlaceholder/LoadingPlaceholder"
import Icon from "../Icon/Icon"
import { ReactComponent as TrashIcon } from "../../icons/trash.svg"
import { navigate } from "../Link/Link"
import useVisibilityState from "../../hooks/useVisibilityState"
import ImageCard from "./ImageCard"
import ObservationMetaCard from "./ObservationMetaCard"
import AlertDialog from "../AlertDialog/AlertDialog"
import useDeleteObservation from "../../hooks/api/observations/useDeleteObservation"
import Checkbox from "../Checkbox/Checkbox"
import DetailsCard from "./DetailsCard"
import usePatchObservation from "../../hooks/api/observations/usePatchObservation"
import LoadingIndicator from "../LoadingIndicator/LoadingIndicator"

export interface PageObservationDetailsProps {
  observationId: string
  studentId: string
  backUrl: string
}
export const PageObservationDetails: FC<PageObservationDetailsProps> = ({
  observationId,
  studentId,
  backUrl,
}) => {
  const deleteDialog = useVisibilityState()
  const { data, isLoading } = useGetObservation(observationId)
  const deleteObservation = useDeleteObservation(observationId, studentId)

  const callDeleteObservationApi = async () => {
    try {
      await deleteObservation.mutate()
      navigate(backUrl)
    } catch (e) {
      Sentry.captureException(e)
    }
  }

  if (isLoading) {
    return <Loading />
  }

  return (
    <Box>
      <ObservationMetaCard observationId={observationId} />

      <DetailsCard
        originalValue={data?.longDesc ?? ""}
        observationId={observationId}
      />

      <ImageCard
        studentId={studentId}
        observationId={observationId}
        originalValue={data?.images ?? []}
      />

      <Flex sx={{ alignItems: "center" }}>
        <VisibleToGuardians
          observationId={observationId}
          originalValue={data?.visibleToGuardians ?? false}
        />
        <Button
          data-cy="delete-observation"
          variant="outline"
          color="danger"
          ml="auto"
          mr={3}
          my={3}
          px={2}
          onClick={deleteDialog.show}
          sx={{ flexShrink: 0 }}
        >
          <Icon as={TrashIcon} fill="danger" />
        </Button>
      </Flex>

      {deleteDialog.visible && (
        <AlertDialog
          title={t`Delete Observation?`}
          onNegativeClick={deleteDialog.hide}
          positiveText="Yes"
          loading={deleteObservation.isLoading}
          body={t`"${
            data?.shortDesc ?? ""
          }" will be permanently deleted. Are you sure?`}
          onPositiveClick={callDeleteObservationApi}
        />
      )}
    </Box>
  )
}

const VisibleToGuardians: FC<{
  observationId: string
  originalValue: boolean
}> = ({ observationId, originalValue }) => {
  const [value, setValue] = useState(originalValue)
  const patchObservation = usePatchObservation(observationId, () =>
    setValue(originalValue)
  )

  const patchVisibleToGuardians = async (visibleToGuardians: boolean) => {
    try {
      setValue(visibleToGuardians)
      await patchObservation.mutateAsync({ visibleToGuardians })
    } catch (e) {
      Sentry.captureException(e)
    }
  }

  return (
    <Flex
      sx={{
        alignItems: "center",
        opacity: patchObservation.isLoading ? 0.5 : 1,
        transition: "opacity 200ms ease-in-out",
      }}
    >
      <Checkbox
        checked={value}
        label={t`Visible to guardians`}
        containerSx={{ ml: [3, 4] }}
        disabled={patchObservation.isLoading}
        onChange={patchVisibleToGuardians}
      />
      {patchObservation.isLoading && <LoadingIndicator />}
    </Flex>
  )
}

const Loading = () => (
  <Box mx={3}>
    <LoadingPlaceholder sx={{ height: 210 }} mb={2} />
    <LoadingPlaceholder sx={{ height: 150 }} mb={2} />
    <LoadingPlaceholder sx={{ height: 92 }} mb={2} />
  </Box>
)

export default PageObservationDetails
