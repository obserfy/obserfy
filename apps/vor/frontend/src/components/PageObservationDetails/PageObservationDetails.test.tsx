import { render } from "../../test-utils"
import { Default } from "./PageObservationDetails.stories"

describe("PageObservationDetails", () => {
  it("should render correctly", () => {
    const { container } = render(<Default />)
    expect(container).toMatchSnapshot()
  })
})
