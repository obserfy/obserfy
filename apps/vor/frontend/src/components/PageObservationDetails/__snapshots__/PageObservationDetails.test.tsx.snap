// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageObservationDetails should render correctly 1`] = `
@keyframes animation-0 {
  0% {
    opacity: 0.05;
  }

  50% {
    opacity: 0.2;
  }

  100% {
    opacity: 0.05;
  }
}

@keyframes animation-0 {
  0% {
    opacity: 0.05;
  }

  50% {
    opacity: 0.2;
  }

  100% {
    opacity: 0.05;
  }
}

@keyframes animation-0 {
  0% {
    opacity: 0.05;
  }

  50% {
    opacity: 0.2;
  }

  100% {
    opacity: 0.05;
  }
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
}

@media screen and (min-width: 40em) {
  .emotion-0 {
    font-size: 20px;
  }
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-left: 16px;
  margin-right: 16px;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-bottom: 8px;
  border-radius: 6px;
  -webkit-animation: 4s infinite animation-0;
  animation: 4s infinite animation-0;
  background-color: #a5a5a5;
  height: 210px;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-bottom: 8px;
  border-radius: 6px;
  -webkit-animation: 4s infinite animation-0;
  animation: 4s infinite animation-0;
  background-color: #a5a5a5;
  height: 150px;
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-bottom: 8px;
  border-radius: 6px;
  -webkit-animation: 4s infinite animation-0;
  animation: 4s infinite animation-0;
  background-color: #a5a5a5;
  height: 92px;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      />
      <div
        class="emotion-3"
      />
      <div
        class="emotion-4"
      />
    </div>
  </div>
</div>
`;
